be/src/exec/range_router.cpp

```diff
@@ -0,0 +1,311 @@
+#include "fmt/format.h"
+#include "runtime/descriptors.h"
+#include "storage/types.h"
+
+namespace starrocks {
+
+Status RangeRouter::init(const std::vector<TTabletRange>& tablet_ranges, size_t num_columns) {
+    if (!_lower_boundaries.empty()) {
+        DCHECK((_lower_boundaries[0] == nullptr) || (_lower_boundaries[0]->size() + 1 == _lower_bound_inclusive.size()));
+        return Status::OK();
+    }
+
+    if (tablet_ranges.empty()) {
+        return Status::InternalError("no tablet ranges for RangeRouter init");
+    }
+
+    const size_t num_ranges = tablet_ranges.size();
+    _lower_boundaries.resize(num_columns);
+    _lower_bound_inclusive.resize(num_ranges, false);
+
+    RETURN_IF_ERROR(_validate_range(tablet_ranges, num_columns));
+    // (-inf, +inf) range, fast return
+    if (num_ranges == 1) {
+        return Status::OK();
+    }
+
+    // Infer TypeDescriptor for each range-distributed column from tablet ranges.
+    // Caller should guarantee that for every column for range distribution, at least one
+    // tablet range provides a concrete type information for lower bound or upper bound.
+    std::vector<std::unique_ptr<TypeDescriptor>> type_descs(num_columns);
+    for (size_t i = 0; i < num_columns; ++i) {
+        for (const auto& tablet_range : tablet_ranges) {
+            const TTuple* lower = tablet_range.__isset.lower_bound ? &tablet_range.lower_bound : nullptr;
+            const TTuple* upper = tablet_range.__isset.upper_bound ? &tablet_range.upper_bound : nullptr;
+            DCHECK(lower != nullptr || upper != nullptr);
+
+            if (lower != nullptr) {
+                DCHECK(i < lower->values.size());
+                type_descs[i] = std::make_unique<TypeDescriptor>(TypeDescriptor::from_thrift(lower->values[i].type));
+                break;
+            }
+
+            if (upper != nullptr) {
+                DCHECK(i < upper->values.size());
+                type_descs[i] = std::make_unique<TypeDescriptor>(TypeDescriptor::from_thrift(upper->values[i].type));
+                break;
+            }
+        }
+    }
+
+    auto parse_bound = [&](const TTuple& bound, std::vector<ColumnPtr>& boundary_columns) {
+        for (size_t i = 0; i < num_columns; ++i) {
+            const auto& type_desc = type_descs[i];
+            DCHECK(type_desc != nullptr);
+            ColumnPtr& column = boundary_columns[i];
+            if (!column) {
+                column = ColumnHelper::create_column(*type_desc, false);
+                column->reserve(num_ranges);
+            }
+
+            const auto& value = bound.values[i];
+            Datum datum;
+            std::string value_str;
+            auto type_info = get_type_info(type_desc->type, type_desc->precision, type_desc->scale);
+            if (value.__isset.string_value) {
+                value_str = value.string_value;
+            } else if (value.__isset.long_value) {
+                value_str = std::to_string(value.long_value);
```

Comment:
so only need string_value in proto?

Reply:
Yes, i have refactor the code and remove the long value part